### PR TITLE
simplified installer (8/9): support FDO

### PR DESCRIPTION
**Pull request by runcom**
This patch also introduces a new customization manufacturing_service_url
to allow users to customize it.

~~BLOCKER: we haven't packaged FIDO yet, that's why this is a draft for 8.6/9.0~~
NICE2HAVE (we don't support this atm in FDO itself): FDO stage https://github.com/osbuild/osbuild/pull/857

Signed-off-by: Antonio Murdaca <runcom@linux.com>

This pull request includes:

- [ ] adequate testing for the new functionality or fixed issue
- [ ] adequate documentation informing people about the change such as
  - [ ] create a file in [news/unreleased](https://github.com/osbuild/osbuild-composer/tree/main/docs/news/unreleased) directory if this change should be mentioned in the release news
  - [ ] submit a PR for the [guides](https://github.com/osbuild/guides) repository if this PR changed any behavior described there: https://www.osbuild.org/guides/

### CODE REVIEW

**Comment by runcom** [comment]
moved everything to 8.6

**Comment by runcom** [comment]
enabled fido for RHEL 9 as well - this now depends on https://github.com/osbuild/osbuild-composer/pull/2015 (I'll rebase once it merges)

**Comment by henrywang** [comment]
@runcom PR #2015 merged. It's time to work on this one. :1234:

**Comment by henrywang** [comment]
@runcom we need https://github.com/osbuild/osbuild/pull/857 merged. And we can pin osbuild to that commit.

**Comment by runcom** [comment]
> @runcom we need [osbuild/osbuild#857](https://github.com/osbuild/osbuild/pull/857) merged. And we can pin osbuild to that commit.

yeah, although we still need to package FDO in rhel so this is currently blocked on that anyway

File: internal/distro/rhel90/distro.go
Comment (runcom) [inline internal/distro/rhel90/distro.go:471]: we need a test plan that:

- configures and brings up a manufacturing server (via container perhaps)
- specify the FDO manufacturing server url in the blueprint
- specify either insecure or hash to verify the connection (test both)
- test installation

File: internal/distro/rhel90/distro.go
Comment (runcom) [inline internal/distro/rhel90/distro.go:578]: we'd also need an onboarding test plan like:

- configure and bring up a rendezvous and onboarding server
- configure and bring up a manufacturing server
- configure SSH in the onboarding server server info
- install
-  verify the ssh is there

File: internal/distro/rhel90/distro.go
Comment (runcom) [reply internal/distro/rhel90/distro.go:471]: probably better to validate the full flow as per my comment below

File: internal/distro/rhel90/stage_options.go
Comment (runcom) [inline internal/distro/rhel90/stage_options.go:325]: FDO can't be `nil` as we make sure it's not, drop this

File: internal/distro/rhel90/stage_options.go
Comment (runcom) [inline internal/distro/rhel90/stage_options.go:326]: this can't be nil either

File: internal/blueprint/customizations.go
Comment (runcom) [inline internal/blueprint/customizations.go:28]: this should be just true or false, validate if present

File: internal/distro/rhel86/distro.go
Comment (runcom) [inline internal/distro/rhel86/distro.go:479]: validate it's either true or false after the empty check

File: internal/blueprint/customizations.go
Comment (runcom) [reply internal/blueprint/customizations.go:28]: nevermind, setting this just flip insecure

**Comment by ondrejbudai** [comment]
FTR this needs regenerated test cases.

**Comment by henrywang** [comment]
Thanks @ondrejbudai. Yes, @yih-redhat is working with @runcom on test.

File: internal/distro/rhel90/package_sets.go
Comment (nullr0ute) [inline internal/distro/rhel90/package_sets.go:118]: For reference, and this is perfectly fine, the reason the ia32 grub2/shim gets pulled in by default there's been a number of very bad devices that are 64bit but with 32bit firmware and those are needed to boot it and then the kernel runs 64 bit. I doubt we'll need to ever support such a device but I'm adding this comment for reference.

File: internal/distro/rhel90/distro.go
Comment (runcom) [reply internal/distro/rhel90/distro.go:475]: Yi is testing this case with insecure - we can evolve the test for hash check later on

**Comment by jrusz** [comment]
Hi @runcom can you please rebase before you push more changes? There was an issue in CI that's causing failed jobs to wait for timeout which then wastes CI resources, thanks!

**Comment by runcom** [comment]
> Hi @runcom can you please rebase before you push more changes? There was an issue in CI that's causing failed jobs to wait for timeout which then wastes CI resources, thanks!

done!

**Comment by runcom** [comment]
This is blocked on https://github.com/osbuild/osbuild-composer/pull/2219 and https://github.com/osbuild/osbuild-composer/pull/2196 and new snapshots for rpmrepo

File: internal/distro/rhel86/distro.go
Comment (runcom) [inline internal/distro/rhel86/distro.go:635]: we have a follow up card for this so we can get this in and I'll clean up once that's done (this allows us to test the flow as having a preset)

**Review by achilleas-k — DISMISSED** [review]
Wonderful.  Thanks you for this!

**Review by achilleas-k — APPROVED** [review]

**Comment by teg** [comment]
This needs a rebase to fix `ERROR! Invalid callback for stdout specified: yaml`.

**Comment by achilleas-k** [comment]
Fixing up the edge manifests.

**Comment by runcom** [comment]
Can we kick tests again? Afaict it should be good now :)

**Comment by runcom** [comment]
rebased

**Comment by gicmo** [comment]
I think the rebase picked up the container issue #2330: `Error: parsing image configuration: parsing time "\"2022-02-22T10:59:43+00:00Z\"" as "\"2006-01-02T15:04:05Z07:00\"": cannot parse "Z\"" as "\""`

**Comment by teg** [comment]
Rebased on top of LVM PR and various CI fixes.

**Comment by runcom** [comment]
The error in 8.6 is this:
```
error: ../../grub-core/loader/i386/efi/linux.c:215:/images/pxeboot/vmlinuz has
invalid signature.
error: ../../grub-core/loader/i386/efi/linux.c:94:you need to load the kernel
first.

Press any key to continue...
```

it seems the kernel we're installing has invalid signature? how so?

**Review by achilleas-k — APPROVED** [review]

File: internal/distro/rhel86/distro.go
Comment (gicmo) [inline internal/distro/rhel86/distro.go:552]: This is quite literally the same error message. Could unify them.

File: internal/distro/rhel86/pipelines.go
Comment (gicmo) [inline internal/distro/rhel86/pipelines.go:779]: We don't have that stage yet.

**Review by gicmo — CHANGES_REQUESTED** [review]
Let's hold off with this until we have the stage in osbuild.

File: internal/distro/rhel86/distro.go
Comment (runcom) [reply internal/distro/rhel86/distro.go]: fixed

File: test/cases/ostree-simplified-installer.sh
Comment (gicmo) [inline test/cases/ostree-simplified-installer.sh:291]: 🙈

File: test/cases/ostree-simplified-installer.sh
Comment (gicmo) [inline test/cases/ostree-simplified-installer.sh:288]: I wonder if we should create a tag and then pin the tag here. We like to test with/against versions of stuff.

Comment (teg) [reply test/cases/ostree-simplified-installer.sh:288]: or even better: just a fixed commit (tags can change)

File: test/cases/ostree-simplified-installer.sh
Comment (runcom) [reply test/cases/ostree-simplified-installer.sh:291]: Walsh would scream about that, I need to create a Bz for it

File: test/cases/ostree-simplified-installer.sh
Comment (runcom) [reply test/cases/ostree-simplified-installer.sh:288]: I’m ok with that ofc - we’re gonna switch the create keys script with a rust cli command in fdo in the next release anyway and we’re probably adding a command to generate configs too (soon), for now I think pinning the commit works

File: test/cases/ostree-simplified-installer.sh
Comment (runcom) [reply test/cases/ostree-simplified-installer.sh:291]: ok we just need to relabel with `:z` pushing a commit to fix after tests run

**Comment by gicmo** [comment]
I will do the adaptions for the fdo stage tomorrow.

**Comment by runcom** [comment]
the simplified installer tests are passing so this should be good now - just the centos-9 koji test failing across the repo anyway. Also, since the FDO version we're shipping today doesn't leverage the new fdo stage, I propose to drop it from this PR if we don't want to merge as is (even if I think we could), this way this PR goes in and QE can start downstream testing - this PR also misses manifests regeneration which I think we'll need at some point

**Comment by runcom** [comment]
(it's amazing finally super green in CI :heart_eyes: )

**Comment by gicmo** [comment]
Manually tested by @runcom, for automated testing we need https://github.com/fedora-iot/fido-device-onboard-rs/pull/225 to be in a fdo server container and additional tests. Will come later.

**Comment by gicmo** [comment]
All green except koji integration on CS9, which is sadly expected, merging.

**Pull request merged**
